clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java

```diff
@@ -634,12 +634,12 @@ public boolean matches(ClientRequest request) {
 
 
     @Test
-    public void testExcludeInternalTopicsConfigOption() { 
+    public void testExcludeInternalTopicsConfigOption() {
         subscriptions.subscribe(Pattern.compile(".*"), rebalanceListener);
 
         metadata.update(TestUtils.singletonCluster(TopicConstants.GROUP_METADATA_TOPIC_NAME, 2), time.milliseconds());
 
-        assertFalse(subscriptions.partitionAssignmentNeeded());
+        assertTrue(subscriptions.partitionAssignmentNeeded());
```

Comment:
I actually tested that scenario with [a simple Java consumer](https://github.com/vahidhashemian/kafka-tests/blob/trunk/src/com/github/vahidhashemian/kafka/samples/ConsumerSubscriptionSemantics.java), subscribed to a regex, and then, in a loop, kept polling every few seconds. When I changed the number of partitions that matched the regex, the rebalance listener of the consumer was not called, and its assigned partitions did not change.

If I comment out the added line `this.needsPartitionAssignment = true;` in `SubscriptionState`, I don't see a rebalance occur and the assigned partitions of the consumer remain the same after a subsequent regex subscription that's supposed to change them.